plugins/states/generational_state_fact.py

```diff
@@ -0,0 +1,249 @@
+    """
+    The legacy state fact is only composed of the state dict.
+    The dict we receive as argument is then the state dict, that we can
+    assign to the state attribute.
+    """
+
+    state: dict
+
+    def get_state(self) -> dict:
+        return self.state
+
+    @classmethod
+    def build_from_state(cls, state: dict) -> "LegacyStateFact":
+        return cls(state=state)
+
+    @classmethod
+    def _convert(cls, state: StateFact) -> typing.Optional["LegacyStateFact"]:
+        """
+        The legacy state fact should not be the desired type for any conversion
+        """
+        return None
+
+
+class GenerationalStateFact(StateFact):
+    @classmethod
+    @abc.abstractmethod
+    def generation(cls) -> str:
+        """
+        Should be implemented by the subclass, and return the generation identifier
+        """
+
+    def _iter(self, **kwargs) -> "pydantic.TupleGenerator":
+        """
+        We overwrite the _iter method simply to add our generation marker to the
+        generated dict or json payload.
+        """
+        for x in super()._iter(**kwargs):
+            yield x
+
+        yield STATE_DICT_GENERATION_MARKER, self.generation()
+
+    @classmethod
+    def build_from_state(cls: typing.Type[GSF], state: dict) -> GSF:
+        state_generation = state[STATE_DICT_GENERATION_MARKER]
+        if not state_generation == cls.generation():
+            # Actively check that the state dict is of the correct generation
+            raise ValueError(
+                f"Unexpected generation value: {state_generation} != {cls.generation()}"
+            )
+
+        return cls(**state)
+
+
+class AlbatrossGenerationStateFact(GenerationalStateFact):
+    """
+    This is the first iteration on a generational approach for state dicts.
+    The generation is named "Albatross" because it is an animal starting with
+    letter A.  It is expected for the following generation to respect this
+    scheme and pick another animal with the next letter coming in the alphabetic
+    order.
+    """
+
+    state: dict
+    created_at: datetime.datetime
+    updated_at: datetime.datetime
+    config_hash: str
+
+    def get_state(self) -> dict:
+        return self.state
+
+    @classmethod
+    def generation(cls) -> str:
+        return "Albatross"
+
+    @classmethod
+    def _convert(
+        cls, state: StateFact
+    ) -> typing.Optional["AlbatrossGenerationStateFact"]:
+        """
+        The only source state fact we know how to recover from is the legacy one.  So we
+        "convert" the input state fact to the legacy one and then upgrade it.
+        """
+        legacy_state_fact = LegacyStateFact.convert(state)
+        return AlbatrossGenerationStateFact(
+            state=legacy_state_fact.state,
+            created_at=datetime.datetime.now().astimezone(),  # Make our date timezone-aware
+            updated_at=datetime.datetime.now().astimezone(),  # Make our date timezone-aware
+            config_hash="",
```

Comment:
If the config hash is empty, the state will be rejected by this:

```  if (
        generational_state_fact.AlbatrossGenerationStateFact.convert(
            previous_state_wrapper
        ).config_hash
        != current_config_hash
    ):
        # The config and the state we have in cache are out of sync, it is
        # unsafe to use, so we return an Unknown
        return Unknown(source=resource)
        
```

Doesn't this defeat the purpose of converting?

Reply:
No, the idea is of the dict hash is that it guarantees the state dict is safe to use in the current model (the state corresponds to the config in the model).  When we do the conversion, we have no way of making sure of that, so we produce a hash that will be invalid, and will trigger unknowns, until the next deployment of the resource, during which the new state dict will be saved.

Reply:
Why have conversion at all then?

Reply:
So that we don't need to handle every possible generation of state dict explicitly anywhere we receive a state from the api.  We simply convert the state dict to the latest generation, and if it is usable, we use it, otherwise it will be usable on the next compile.